# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxPlatformMac.cpp
# Commit: 8928abb7996b
# Full Hash: 8928abb7996b8d778fccbedb6b2d59f38b25b90b
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2021-05-11 09:33:39
# Description:
#   Bug 1708821 - On pre-10.15 macOS versions, don't do off-main-thread font registration in content processes. r=lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D114484
# ==============================================================================

diff -r 8537afd2dcaa -r 8928abb7996b gfx/thebes/gfxPlatformMac.cpp
--- a/gfx/thebes/gfxPlatformMac.cpp	Mon May 10 20:49:31 2021 +0000
+++ b/gfx/thebes/gfxPlatformMac.cpp	Mon May 10 21:04:53 2021 +0000
@@ -162,14 +162,23 @@
    our font list. */
 /* static */
 void gfxPlatformMac::RegisterSupplementalFonts() {
-  // On Catalina+, it appears to be sufficient to activate fonts in the parent
-  // process; they are then also usable in child processes. But on pre-Catalina
-  // systems we need to explicitly activate them in each child process (per bug
-  // 1704273).
-  if (XRE_IsParentProcess() || !nsCocoaFeatures::OnCatalinaOrLater()) {
+  if (XRE_IsParentProcess()) {
     sFontRegistrationThread = PR_CreateThread(
         PR_USER_THREAD, FontRegistrationCallback, nullptr, PR_PRIORITY_NORMAL,
         PR_GLOBAL_THREAD, PR_JOINABLE_THREAD, 0);
+  } else if (!nsCocoaFeatures::OnCatalinaOrLater()) {
+    // On Catalina+, it appears to be sufficient to activate fonts in the
+    // parent process; they are then also usable in child processes. But on
+    // pre-Catalina systems we need to explicitly activate them in each child
+    // process (per bug 1704273).
+    //
+    // But at least on 10.14 (Mojave), doing font registration on a separate
+    // thread in the content process seems crashy (bug 1708821), despite the
+    // CTFontManager.h header claiming that it's thread-safe. So we just do it
+    // immediately on the main thread, and accept the startup-time hit (sigh).
+    for (const auto& dir : kLangFontsDirs) {
+      ActivateFontsFromDir(dir);
+    }
   }
 }
 
